libs/types-rs/src/bmd/completed_ballot.rs

```diff
@@ -0,0 +1,186 @@
+        &self,
+        w: &mut W,
+        election: &Self::Context,
+    ) -> Result<(), Self::Error>
+    where
+        Self: Sized,
+    {
+        w.write_bytes(PRELUDE)?;
+        w.write_bytes(&self.ballot_hash)?;
+
+        let precinct_index = election
+            .precinct_index(&self.precinct_id)
+            .ok_or_else(|| Error::InvalidPrecinctId(self.precinct_id.clone()))?;
+        w.build(&precinct_index)?;
+
+        let ballot_style_index = election
+            .ballot_style_index(&self.ballot_style_id)
+            .ok_or_else(|| Error::InvalidBallotStyleId(self.ballot_style_id.clone()))?;
+        w.build(&ballot_style_index)?;
+
+        let ballot_style = election
+            .ballot_styles
+            .get(ballot_style_index.get() as usize)
+            .ok_or_else(|| Error::InvalidBallotStyleIndex {
+                index: ballot_style_index.get() as usize,
+                count: election.ballot_styles.len(),
+            })?;
+
+        w.write_bit(self.is_test_mode)?;
+        w.build(&self.ballot_type)?;
+
+        let contests = election.contests_in(ballot_style);
+
+        match self.ballot_audit_id {
+            Some(ref ballot_audit_id) => {
+                w.write_bit(true)?;
+                let Ok(ballot_audit_id_length) = u8::try_from(ballot_audit_id.len()) else {
+                    return Err(Error::InvalidBallotAuditId(ballot_audit_id.clone()));
+                };
+                let Some(ballot_audit_id_length) = BallotAuditIdLength::new(ballot_audit_id_length)
+                else {
+                    return Err(Error::InvalidBallotAuditId(ballot_audit_id.clone()));
+                };
+                w.build(&ballot_audit_id_length)?;
+                w.write_bytes(ballot_audit_id.as_bytes())?;
+            }
+
+            None => w.write_bit(false)?,
+        }
+
+        // write roll call
+        for contest in &contests {
+            w.write_bit(self.votes.iter().any(|contest_vote| match contest_vote {
+                ContestVote::Candidate { contest_id, votes } if contest_id == contest.id() => {
+                    !votes.is_empty()
+                }
+                ContestVote::YesNo { contest_id, .. } if contest_id == contest.id() => true,
+                _ => false,
+            }))?;
+        }
+
+        // write vote data
+        for contest in &contests {
+            let contest_votes = self
+                .votes
+                .iter()
+                .filter(|contest_vote| contest_vote.contest_id() == contest.id())
+                .collect::<Vec<_>>();
+
+            let contest_vote = match contest_votes[..] {
+                [] => continue,
+                [_, _, ..] => {
+                    return Err(Error::InvalidVotes {
+                        message: "Too many ContestVote records for contest".to_owned(),
+                    })
+                }
+                [contest_vote] => contest_vote,
+            };
```

Comment:
Any reasoning behind representing `votes` as a vector rather than a mapping (like we do in the corresponding TS code)? This guard in particular would be unnecessary if there was a structural constraint that each contest could only have one vote record

Reply:
d36c8e5d